storage/range.go

```diff
@@ -1759,27 +1760,64 @@ func (r *Range) AdminMerge(args *proto.AdminMergeRequest, reply *proto.AdminMerg
 	}
 }
 
-// InternalChangeReplicas is called after a raft configuration change has committed.
-func (r *Range) InternalChangeReplicas(batch engine.Engine,
-	req *proto.InternalChangeReplicasRequest,
-	reply *proto.InternalChangeReplicasResponse) {
-	r.Lock()
-	defer r.Unlock()
-	// Apply the committed membership change to r.Desc.
-	// TODO(bdarnell): update the range addressing records too (atomically).
-	// Need to either combine this with EndTransaction or queue up a transaction on
-	// the meta records.
-	r.Desc.Replicas = r.Desc.Replicas[:0]
-	for _, n := range req.Nodes {
-		nodeID, storeID := decodeRaftNodeID(multiraft.NodeID(n))
-		r.Desc.Replicas = append(r.Desc.Replicas, proto.Replica{
-			NodeID:  nodeID,
-			StoreID: storeID,
-			// TODO(bdarnell): Set attributes. Copy from gossip or pass through request?
-		})
+// ChangeReplicas adds or removes a replica of a range. The change is performed
+// in a distributed transaction and takes effect when that transaction is committed.
+// When removing a replica, only the NodeID and StoreID fields of the Replica are used.
+func (r *Range) ChangeReplicas(changeType proto.ReplicaChangeType, replica proto.Replica) error {
+	// Only allow a single change per range at a time.
+	if !atomic.CompareAndSwapInt32(&r.metaLock, int32(0), int32(1)) {
+		return util.Errorf("range %d metadata locked", r.Desc.RaftID)
+	}
+	defer func() { atomic.StoreInt32(&r.metaLock, int32(0)) }()
+
+	// Validate the request and prepare the new descriptor.
+	updatedDesc := *r.Desc
+	found := -1
+	for i, existingRep := range r.Desc.Replicas {
+		if existingRep.NodeID == replica.NodeID && existingRep.StoreID == replica.StoreID {
+			found = i
+			break
+		}
+	}
+	if changeType == proto.ADD_REPLICA {
```

Comment:
Really? Tying configs to replica order seems fragile; I assumed that the zone configs specified attributes of the set as a whole. (my previous commit on this topic reordered Desc.Replicas arbitrarily on any change). This algorithm can work for add-then-remove (assuming we maintain some sort of lock so there are no other changes in between), but it breaks down for remove-then-add, and if we're changing from 3 to 5 replicas or vice versa it gets complicated to maintain the desired ordering without causing unnecessary churn at the raft level (if we add the replica at index 5 before the one at index 4 do we need to create a gap in the list?)